orttraining/orttraining/training_api/utils.cc

```diff
@@ -0,0 +1,68 @@
+#include "core/graph/model.h"
+#include "core/framework/session_state.h"
+#include "core/framework/ort_value.h"
+#include "core/framework/tensor.h"
+#include "core/framework/allocator.h"
+
+#include "orttraining/training_api/include/utils.h"
+
+namespace onnxruntime {
+namespace training {
+namespace api {
+
+void GetGraphInputOutputNames(const Graph& graph,
+                              std::vector<std::string>& input_names,
+                              std::vector<std::string>& output_names) {
+  auto inputs = graph.GetInputs();
+  auto outputs = graph.GetOutputs();
+
+  auto get_names = [&](const std::vector<const NodeArg*>& node_args, std::vector<std::string>& names) {
+    for (const auto* arg : node_args) {
+      names.push_back(arg->Name());
+    }
+  };
+
+  get_names(inputs, input_names);
+  get_names(outputs, output_names);
+}
+
+bool GetParamNameFromSuffix(const std::string& name, const std::string& suffix, std::string& param_name) {
+  bool endswith = std::equal(suffix.rbegin(), suffix.rend(), name.rbegin());
+  if (endswith) {
+    param_name = name.substr(0, name.length() - suffix.length());
+    return true;
+  } else {
+    param_name = "";
+    return false;
+  }
+}
+
+bool GetParamNameFromGradient(const std::string& grad_name, std::string& param_name) {
+  for (auto& suffix : GRAD_SUFFIX) {
+    if (GetParamNameFromSuffix(grad_name, suffix, param_name)) {
```

Comment:
This is tricky... we loop the suffix and do the match. ..

Reply:
yes because the frontend tooling is generating different gradient names when I tested. We can possible consolidate this with the frontend tooling PR to use the same source for the name. will add TODO here